crates/database/src/settlement_observations.rs

```diff
@@ -0,0 +1,84 @@
+use {crate::events::EventIndex, bigdecimal::BigDecimal, sqlx::PgConnection};
+
+#[derive(Debug, PartialEq, sqlx::FromRow)]
+pub struct Observation {
+    pub gas_used: BigDecimal,
+    pub effective_gas_price: BigDecimal,
+    pub surplus: BigDecimal,
+    pub fee: BigDecimal,
+    pub block_number: i64,
+    pub log_index: i64,
+}
+
+pub async fn upsert(
+    ex: &mut PgConnection,
+    event: &EventIndex,
+    gas_used: BigDecimal,
+    effective_gas_price: BigDecimal,
+    surplus: BigDecimal,
+    fee: BigDecimal,
+) -> Result<(), sqlx::Error> {
+    const QUERY: &str = r#"
+    INSERT INTO settlement_observations (gas_used, effective_gas_price, surplus, fee, block_number, log_index) 
+    VALUES ($1, $2, $3, $4, $5, $6)
+    ON CONFLICT (block_number, log_index) DO UPDATE SET gas_used = $1, effective_gas_price = $2, surplus = $3, fee = $4;
+    "#;
+    sqlx::query(QUERY)
+        .bind(gas_used)
+        .bind(effective_gas_price)
+        .bind(surplus)
+        .bind(fee)
+        .bind(event.block_number)
+        .bind(event.log_index)
+        .execute(ex)
+        .await?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, sqlx::Connection};
+
+    // helper function to make roundtrip possible
+    pub async fn fetch(
+        ex: &mut PgConnection,
+        event: &EventIndex,
+    ) -> Result<Option<Observation>, sqlx::Error> {
```

Comment:
I guess there is no need for this in the actual program so it lives here until its needed?

Reply:
Yes, it's needed for tests only (at least so far, will see how the data will be used but I predict it's gonna be joined with something)